libraries/plugins/custom_operations/include/graphene/custom_operations/custom_operations.hpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2019 oxarbitrage and contributors.
+ *
+ * The MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#pragma once
+
+#include <graphene/chain/database.hpp>
+
+#include "custom_objects.hpp"
+
+namespace graphene { namespace custom_operations {
+
+using namespace std;
+using graphene::protocol::account_id_type;
+
+struct account_storage_map : chain::base_operation
+{
+   struct ext
+   {
+      optional<bool> remove;
+      optional<string> catalog;
+      optional<flat_map<string, string>> key_values;
+   };
+
+   graphene::protocol::extension<ext> extensions;
```

Comment:
Why so complicated? Just make them normal non-optional fields, that saves you from having to check validity later.

Reply:
The `extension/optional` structure was for the old `account_contact_operation` which had lots of fields but not all fields would be used by clients. Here the new operation use a map, thus no need to use the `extension/optional` structure.

Reply:
By the way, if we use `variant` for storage, better use `variant` in the operation as well.

Reply:
If we use `string` in the operations we can then convert to variant in the evaluator before inserting while we reject if the value field is not valid json. It will save us from having a dedicated api for clients to convert json string to variant.

Reply:
If the clients are unable to handle `variant`s, I think it's better to save the strings as strings directly. Converting in/out would probably produce different results which would lead to confusion.

Reply:
Removed the extensions by normal fields at https://github.com/bitshares/bitshares-core/pull/1926/commits/6a12446b958d3a7ff26edbe1372a9cb74f65550b